src/backend/storage/data_table.cpp

```diff
@@ -458,37 +526,17 @@ oid_t DataTable::AddDefaultTileGroup() {
 
   LOG_TRACE("Trying to add a tile group ");
   {
-    std::lock_guard<std::mutex> lock(table_mutex);
-
-    // Check if we actually need to allocate a tile group
-
-    // (A) no tile groups in table
-    if (tile_groups.empty()) {
-      LOG_TRACE("Added first tile group ");
-      tile_groups.push_back(tile_group->GetTileGroupId());
-      // add tile group metadata in locator
-      catalog::Manager::GetInstance().AddTileGroup(tile_group_id, tile_group);
-      LOG_TRACE("Recording tile group : %lu ", tile_group_id);
-      return tile_group_id;
-    }
-
-    // (B) no slots in last tile group in table
-    auto last_tile_group_offset = GetTileGroupCount() - 1;
-    auto last_tile_group = GetTileGroup(last_tile_group_offset);
-
-    oid_t active_tuple_count = last_tile_group->GetNextTupleSlot();
-    oid_t allocated_tuple_count = last_tile_group->GetAllocatedTupleCount();
-    if (active_tuple_count < allocated_tuple_count) {
-      LOG_TRACE("Slot exists in last tile group :: %lu %lu ",
-                active_tuple_count, allocated_tuple_count);
-      return INVALID_OID;
-    }
-
     LOG_TRACE("Added a tile group ");
-    tile_groups.push_back(tile_group->GetTileGroupId());
+    tile_groups_.push_back(tile_group->GetTileGroupId());
 
     // add tile group metadata in locator
     catalog::Manager::GetInstance().AddTileGroup(tile_group_id, tile_group);
+    
+    // we must guarantee that the compiler always add tile group before adding tile_group_count_.
+    COMPILER_MEMORY_FENCE;
```

Comment:
We should move these things to a separate header file that contains all the architecture specific stuff.